pq-crypto/s2n_kyber_512_evp.c

```diff
@@ -20,57 +20,63 @@
 
 #include "error/s2n_errno.h"
 #include "tls/s2n_kem.h"
+#include "utils/s2n_safety.h"
 #include "utils/s2n_safety_macros.h"
 
 #if defined(S2N_LIBCRYPTO_SUPPORTS_KYBER512)
-int s2n_kyber_512_evp_generate_keypair(uint8_t *public_key, uint8_t *private_key) {
-    EVP_PKEY_CTX *kyber_pkey_ctx = EVP_PKEY_CTX_new_id(EVP_PKEY_KYBER512, NULL);
+
+DEFINE_POINTER_CLEANUP_FUNC(EVP_PKEY *, EVP_PKEY_free);
+DEFINE_POINTER_CLEANUP_FUNC(EVP_PKEY_CTX *, EVP_PKEY_CTX_free);
+
+int s2n_kyber_512_evp_generate_keypair(uint8_t *public_key, uint8_t *secret_key) {
+    DEFER_CLEANUP(EVP_PKEY_CTX *kyber_pkey_ctx = EVP_PKEY_CTX_new_id(EVP_PKEY_KEM, NULL), EVP_PKEY_CTX_free_pointer);
     POSIX_GUARD_PTR(kyber_pkey_ctx);
+    POSIX_ENSURE(EVP_PKEY_CTX_kem_set_params(kyber_pkey_ctx, NID_KYBER512_R3), S2N_FAILURE);
     POSIX_ENSURE(EVP_PKEY_keygen_init(kyber_pkey_ctx), S2N_FAILURE);
 
-    EVP_PKEY *kyber_pkey = NULL;
+    DEFER_CLEANUP(EVP_PKEY *kyber_pkey = NULL, EVP_PKEY_free_pointer);
     POSIX_ENSURE(EVP_PKEY_keygen(kyber_pkey_ctx, &kyber_pkey), S2N_FAILURE);
```

Comment:
I'm not sure how I didn't notice this before, but this is not how you use [POSIX_ENSURE](https://github.com/aws/s2n-tls/blob/main/utils/s2n_safety_macros.h#L209-L214). S2N_FAILURE is not an error code, it's a generic failure return value. You need to set a valid error code on failure. See [s2n_errno.h](https://github.com/aws/s2n-tls/blob/main/error/s2n_errno.h) / [s2n_errno.c](https://github.com/aws/s2n-tls/blob/main/error/s2n_errno.c).

Can you fix the rest of the file?

Reply:
sure. we don't expect any of these to fail, so i went with S2N_ERR_INTERNAL_LIBCRYPTO_ERROR for all of them.

https://github.com/aws/s2n-tls/blob/main/error/s2n_errno.h#L306